File: smart_home_comm_v1/main/src/ring_buffer.cpp
#include "../hdr/ring_buffer.h"

volatile int head = 0; // write
volatile int tail = 0; // read
const int max_buffer_size = 32;
volatile unsigned char cbuffer[max_buffer_size];

void init_cbuffer(void)
{
  head = 0;
  tail = 0;
}

void cbuffer_write(unsigned char data)
{
  int nextPos = head + 1;
  if (nextPos == max_buffer_size)  // On the next write the max buffer size is reached - Overwrite of the buffer is required
  {
    nextPos = 0;
    //PORTB ^= (1<<7);
  }
  if (nextPos != tail) // Buffer not full on next write
  {
    cbuffer[head] = data;
    head = nextPos;
  }
  else {
    // Buffer full on next write ((current head + 1) == current tail)
  }
}

unsigned char cbuffer_read(void)
{
  unsigned char temp = '\0';
  int nextPos = tail + 1;
  if (tail != head) // If the buffer is not empty
  {
      if (nextPos == max_buffer_size) // Prevent reading from outside buffer.
      {
        nextPos = 0;
      }
      temp = cbuffer[tail];
      tail = nextPos;
  }
  else {
    // Buffer is currently empty (current tail == current head)
  }
  return temp;
}

bool cbuffer_data_available(void)
{
  return ((head != tail)?true:false);
}

File: smart_home_comm_v1/ring_buffer.py


class RingBuffer:
    def __init__(self, max_size):
        self._max_size = max_size
        self._head = 0
        self._tail = 0
        self._buffer = [0] * self._max_size
    
    def write(self, data):
        next_pos = self._head + 1
        if (next_pos == self._max_size):
            next_pos = 0
        if (next_pos != self._tail):
            self._buffer[self._head] = data
            print(f"Write: Head Index= {self._head}, Value= {data}, Next Head Pos= {next_pos}")
            self._head = next_pos
        else:
            print("Buffer is full on next write")

    def read(self):
        temp = 0
        next_pos = self._tail + 1
        if (self._tail != self._head):
            if (next_pos == self._max_size):
                next_pos = 0
            temp = self._buffer[self._tail]
            print(f"Read: Tail Index= {self._tail}, Value= {temp}, Next Head Pos= {next_pos}")
            self._tail = next_pos
        else:
            print("Buffer is empty")
        return temp

    def data_available(self):
        available = False
        if (self._head != self._tail):
            available = True
        return available




""" 
#################### TEST CODE #################
def main():
    ring = RingBuffer(10)

    for i in range(1, 7):
        ring.write(i)

    print("------------------------------------------------------------------------")
    
    if (ring.data_available()):
        for j in range(1, 3):
            x = ring.read()
            

    print("------------------------------------------------------------------------")

    if (ring.data_available()):
        for j in range(1, 5):
            y = ring.read()

    print("------------------------------------------------------------------------")

    for i in range(1, 7):
        ring.write(i)



    if (ring.data_available()):
        for j in range(1, 12):
            y = ring.read()
    
    for i in range(15):
        ring.write(i)

main() 
"""